explorer/parse_and_execute_test.cpp

```diff
@@ -0,0 +1,67 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "common/check.h"
+#include "common/error.h"
+#include "explorer/interpreter/exec_program.h"
+#include "explorer/interpreter/trace_stream.h"
+#include "explorer/syntax/parse.h"
+#include "explorer/syntax/prelude.h"
+#include "gtest/gtest.h"
+
+namespace Carbon::Testing {
+namespace {
+
+class ParseAndExecuteTest : public ::testing::Test {
+ protected:
+  auto ParseAndExecute(const std::string& source,
+                       testing::Matcher<std::string> err_matcher) -> void {
+    auto err = ParseAndExecuteImpl(source);
+    ASSERT_FALSE(err.ok());
+    EXPECT_THAT(err.error().message(), err_matcher);
+  }
+
+  auto ParseAndExecuteImpl(const std::string& source) -> ErrorOr<int> {
+    Arena arena;
+    CARBON_ASSIGN_OR_RETURN(AST ast,
+                            ParseFromString(&arena, "test.carbon", source,
+                                            /*parser_debug=*/false));
+
+    AddPrelude("explorer/data/prelude.carbon", &arena, &ast.declarations,
+               &ast.num_prelude_declarations);
+    TraceStream trace_stream;
+
+    // Use llvm::nulls() to suppress output from the Print intrinsic.
+    CARBON_ASSIGN_OR_RETURN(
+        ast, AnalyzeProgram(&arena, ast, &trace_stream, &llvm::nulls()));
+    return ExecProgram(&arena, ast, &trace_stream, &llvm::nulls());
+  }
+};
```

Comment:
This unit test will add some maintenance overhead for future refactorings of explorer internals; it might be cheaper for overall maintenance to either check in a script to generate the test below or just create and check in the ~2000-line test as a normal FileCheck test. But I don't feel strongly about this. Feel free to land this either way.

Reply:
Went a step further and also merged in main.cpp logic, so now everything's more uniform.